src/python/pants/core/goals/test.py

```diff
@@ -208,6 +281,90 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
     __test__ = False
 
 
+_TestFieldSetT = TypeVar("_TestFieldSetT", bound=TestFieldSet)
+
+
+@union
+class TestRequest:
+    """Base class for plugin types wanting to be run as part of `test`.
+
+    Plugins should define a new type which subclasses this type, and set the
+    appropriate class variables.
+    E.g.
+        class DryCleaningRequest(TestRequest):
+            tool_subsystem = DryCleaningSubsystem
+            field_set_type = DryCleaningFieldSet
+
+    Then register the rules which tell Pants about your plugin.
+    E.g.
+        def rules():
+            return [
+                *collect_rules(),
+                *DryCleaningRequest.rules(),
+            ]
+    """
+
+    tool_subsystem: ClassVar[type[SkippableSubsystem]]
+    field_set_type: ClassVar[type[TestFieldSet]]
+    partitioner_type: ClassVar[PartitionerType] = PartitionerType.DEFAULT_ONE_PARTITION_PER_INPUT
+
+    __test__ = False
+
+    @classproperty
+    def tool_name(cls) -> str:
+        return cls.tool_subsystem.options_scope
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    class PartitionRequest(_PartitionFieldSetsRequestBase[_TestFieldSetT]):
+        pass
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    class Batch(_BatchBase[PartitionElementT, PartitionMetadataT]):
+        pass
+
+    @memoized_classproperty
+    def _one_input_per_batch_compatibility_rules(cls) -> Iterable:
```

Comment:
Is this permanent? Or planning to remove eventually

Reply:
:man_shrugging: on the asserts, we can hide that behind a helper method.

Reply:
Would it be extra boilerplate? Like I said I wouldn't be opposed to a helper method on the batch class that does the assertion and returns the one result.

Reply:
I also think we should just rip the bandaid now. We already did for `fmt` and `lint`. :smile:

Reply:
Agreed, will update here

Reply:
Pushed a commit removing these + dealing with the knock-on effects